### PR TITLE
Keyboard Shortcuts: MK2

### LINKED ISSUE #816: Keyboard Shortcuts
### Details
Support keyboard navigation and shortcuts.

- [ ] Update CMD + T behavior not to cycle thru states
- [ ] Review scroll position restoration mechanism
- [ ] While in editor, `Cmd + Enter` brings you back to the notes list
- [ ] Add shortcuts
- [ ] Support RTL
- [ ] Add tracks

Action | Shortcut
-|-
Focus search field | Cmd + Shift + F
Focus Mode|Cmd + Option + F
Jump to next match in note | Cmd + G
Jump to previous match in note | Cmd + Shift + G
Toggle editing Content and Tags|Cmd + Shift + Y
Show / Collapse Tags List|Cmd + T
Toggle note list | Cmd + Shift + L
Create new note |  Cmd + N
Toggle markdown preview | Cmd + Shift + P
Insert checklist | Cmd + Shift + C

### PR DESCRIPTION
**Pull request by eshurakov**
### Details
In this PR:
* Cleanup metadata cache (on logout or when the app is completely closed)
* Arrow / Tab navigation skip hidden panels
* Tab navigation between panels
* Update list selection color based on currently active panel
* Tracks

Closes: #816

### Screenshots

Screenshot 2021-02-25 at 15 10 11
Screenshot 2021-02-25 at 15 10 14

### Test

- [x] Arrow / Tab navigation skip hidden panels
- [x] Tab navigates between panels Tags -> Note List -> Editor
- [x] When search bar is focused Tab navigates Search Bar -> Note List -> Editor
- [x] Color of selected row depends on whether current panel is active or not

### Review
Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release
`RELEASE-NOTES.txt` was updated in 2304ef6 with:

> Updated keyboard shortcuts #816

### CODE REVIEW

File: Simplenote/NoteListViewController.swift [hunk 1]
```diff
@@ -55,6 +55,23 @@ class NoteListViewController: NSViewController {
...
+            }
+
+            guard oldValue != isActive else {
+                return
+            }
+
+            for row in 0..<tableView.numberOfRows {
+                let rowView = tableView.rowView(atRow: row, makeIfNecessary: false) as? TableRowView
+                rowView?.isActive = isActive
+            }
```
Comment (jleandroperez) [inline Simplenote/NoteListViewController.swift]: Nipticky: Maybe move to a method? (for consistency / cleanup purposes?)

File: Simplenote/NoteListViewController.swift [hunk 2]
```diff
@@ -497,6 +518,15 @@ extension NoteListViewController: SPTableViewDelegate {
         ///     5.  Same as scenario #1, this ends up refreshing the Editor, and invoking `save()`
         ///     6.  Because of the re-entrant `save()` OP, this scenario will also produce an exception
         ///
+
+        /// We need the following code to avoid text editor scroll animation when using keyboard to select a note from the list
+        /// No other methods are working :facepalm:
+        if let window = view.window,
+           NSApp.currentEvent?.type == .some(.keyDown),
+           let event = NSEvent.otherEvent(with: .applicationDefined, location: .zero, modifierFlags: [], timestamp: 0, windowNumber: window.windowNumber, context: nil, subtype: 0, data1: 0, data2: 0) {
+            NSApp.postEvent(event, atStart: true)
```
Comment (jleandroperez) [inline Simplenote/NoteListViewController.swift:529]: We can't have nice things 😂

File: Simplenote/SimplenoteAppDelegate+Swift.swift [hunk 3]
```diff
@@ -454,8 +457,37 @@ extension SimplenoteAppDelegate: NSMenuItemValidation {
     func validateToogleMarkdownPreviewItem(_ item: NSMenuItem) -> Bool {
         noteEditorViewController.validateToogleMarkdownPreviewItem(item)
     }
+
+    /// Updates `active` state of top view controllers based on the current first responder
+    ///
+    func updateActivePanel(with responder: NSResponder) {
+        let viewControllers: [NSResponder] = [tagListViewController, noteListViewController, noteEditorViewController]
+        var responder = responder
+        while true {
```
Comment (jleandroperez) [inline Simplenote/SimplenoteAppDelegate+Swift.swift]: Suggestion (to drop the `while true`)

```
        let viewControllers: [NSResponder] = [tagListViewController, noteListViewController, noteEditorViewController]
        var nextResponder: NSResponder? = responder

        while let currentResponder = nextResponder {
            if viewControllers.contains(currentResponder) {
                tagListViewController.isActive = tagListViewController == nextResponder
                noteListViewController.isActive = noteListViewController == nextResponder
                break
            }

            nextResponder = currentResponder.nextResponder
        }
```

Okay maybe not worth it 😂 

`while true` always makes me freak out a bit (but all of the scenarios are handled so it's all good!

File: Simplenote/TagListViewController+Swift.swift [hunk 4]
```diff
@@ -282,3 +286,17 @@ extension TagListViewController: SPTextFieldDelegate {
...
+
+// MARK: - Appearance
+//
+extension TagListViewController {
+    @objc
+    func refreshTableRowsActiveStatus() {
+        for row in 0..<tableView.numberOfRows {
+            let tableRow = tableView.rowView(atRow: row, makeIfNecessary: false) as? TableRowView
+            tableRow?.isActive = isActive
+        }
```
Comment (jleandroperez) [inline Simplenote/TagListViewController+Swift.swift]: Q: do you think it'd be worthy to move this into `NSTableView+Simplenote` (and deduplicate?)

File: Simplenote/TagListViewController+Swift.swift [hunk 5]
```diff
@@ -282,3 +286,17 @@ extension TagListViewController: SPTextFieldDelegate {
...
+        tableView.reloadSelectedRow()
```
Comment (jleandroperez) [inline Simplenote/TagListViewController+Swift.swift:297]: Out of curiosity, why do we need to reload the selected row?

**Review by jleandroperez — APPROVED** [review]
I've got a corner case!

1. Select multiple notes in the Notes List (shift + arrow!)
2. Press Arrow Right

As a result, the cursor will "get stuck".

Other than that:

✅ Logout effectively nukes the cache
✅ Light / Dark mode "active state" looks great!

Feel free to :shipit: when ready sir!!

File: Simplenote/NoteListViewController.swift (hunk 1 above)
Comment (eshurakov) [reply Simplenote/NoteListViewController.swift]: Done 👍

File: Simplenote/NoteListViewController.swift (hunk 2 above)
Comment (eshurakov) [reply Simplenote/NoteListViewController.swift:529]: Unfortunately :-)

File: Simplenote/SimplenoteAppDelegate+Swift.swift (hunk 3 above)
Comment (eshurakov) [reply Simplenote/SimplenoteAppDelegate+Swift.swift]: Good suggestion, less code :-)

File: Simplenote/TagListViewController+Swift.swift (hunk 4 above)
Comment (eshurakov) [reply Simplenote/TagListViewController+Swift.swift]: Yes, done! Thanks!

File: Simplenote/TagListViewController+Swift.swift (hunk 5 above)
Comment (eshurakov) [reply Simplenote/TagListViewController+Swift.swift:297]: The cell itself also needs to be updated and not only the row

**Comment by eshurakov** [comment]
@jleandroperez I've made some changes to the code, could you take another look when you have time. Thanks!

**Review by jleandroperez — APPROVED** [review]
Looks super solid!! Thank you for fixing all the things!!

:shipit:

**Comment by pachlava** [comment]
[testing] fix/feature verified on MBP 16" Intel | macOS Big Sur 11.2 | Simplenote 2.9 (11223)

**Pull request merged**